### PR TITLE
Multi anchor crop

### PR DESCRIPTION
**Pull request by aaprasad**
As per #32, the major failure mode we're seeing in the model while animal tracking is when nodes disappear causing a shift in visual features and a switch to occur.  We've tried randomly selecting available node, using the pose centroid as the crop, and reducing the crop size but none have worked. Instead here, we add functionality to stack crops from as many nodes along the channel dimension and use that as input into the model.
* we added timm as the backend for the visual encoder to easily enable arbitrary channel numbers
* generalized the positional encoding to work for arbitrary number of boxes
* added an MLP component to project the embedding back to the correct number of spaces
* removed mutable defaults wherever we saw it

## Summary by CodeRabbit

- **New Features**
  - Improved positional embedding handling for multiple anchors.
  - Enhanced visual encoder initialization with more flexible parameters.
  
- **Bug Fixes**
  - Added exception handling for embedding dimension mismatches.
  - Corrected input box shape handling in embeddings.

- **Enhancements**
  - Switched visual encoder's backend to `timm` for better model support.
  - Updated visual encoder to use `torch.nn.LazyLinear` for output layer initialization.

- **Tests**
  - Expanded test coverage for visual encoder and embeddings with new scenarios and parameters.

### CODE REVIEW

File: biogtr/data_structures.py [hunk 1]
```diff
@@ -56,46 +56,58 @@ def __init__(
...
+        if bbox is None:
+            self._bbox = torch.empty(1, 0, 4)
+
+        elif not isinstance(bbox, torch.Tensor):
             self._bbox = torch.tensor(bbox)
+
         else:
             self._bbox = bbox
 
+        if self._bbox.shape[0] and len(self._bbox.shape) == 1:
```
Comment (talmo) [inline biogtr/data_structures.py:68]: Nit: prefer explicit over implicit boolean conversions, especially with numeric types, i.e.:

```suggestion
        if self._bbox.shape[0] == 0 and len(self._bbox.shape) == 1:
```

File: biogtr/data_structures.py [hunk 2]
```diff
@@ -56,46 +56,58 @@ def __init__(
...
+            self._bbox = self._bbox.unsqueeze(0)  # (n_anchors, 4)
+
+        if self._bbox.shape[1] and len(self._bbox.shape) == 2:
```
Comment (talmo) [inline biogtr/data_structures.py:71]: Same as above

File: biogtr/data_structures.py [hunk 3]
```diff
@@ -56,46 +56,58 @@
...
             self._crop = crop
 
-        if len(self._crop.shape) == 2:
-            self._crop = self._crop.unsqueeze(0).unsqueeze(0)
-        elif len(self._crop.shape) == 3:
-            self._crop = self._crop.unsqueeze(0)
+        if len(self._crop.shape) == 2:  # (h, w)
+            self._crop = self._crop.unsqueeze(0)  # (c, h, w)
+        if len(self._crop.shape) == 3:
+            self._crop = self._crop.unsqueeze(0)  # (1, c, h, w)
```
[the remark this answers is not in the document]
Comment (talmo) [reply biogtr/data_structures.py:94]: Agreed: it's much easier to test if you factorize out parts of your constructor into class or static methods, e.g.:

Class-local helper method:
https://github.com/talmolab/sleap-io/blob/e5539864d4f70669af3a5d1dd1556da08711c8a6/sleap_io/io/video.py#L699-L705

Factorized out complex constructor:
https://github.com/talmolab/sleap-io/blob/e5539864d4f70669af3a5d1dd1556da08711c8a6/sleap_io/model/skeleton.py#L106-L118

File: biogtr/data_structures.py [hunk 4]
```diff
@@ -528,8 +543,8 @@ def __init__(
         video_id: int,
         frame_id: int,
         vid_file: str = "",
-        img_shape: ArrayLike = [0, 0, 0],
-        instances: List[Instance] = [],
+        img_shape: ArrayLike = None,
+        instances: List[Instance] = None,
```
Comment (talmo) [inline biogtr/data_structures.py:547]:
```suggestion
        img_shape: ArrayLike | None = None,
        instances: List[Instance] | None = None,
```

or `Optional[...]` if not py >3.10

File: biogtr/models/embedding.py [hunk 5]
```diff
@@ -62,9 +68,16 @@ def __init__(
         self.temperature = temperature
         self.normalize = normalize
         self.scale = scale
+        self.n_points = n_points
+
         if self.normalize and self.scale is None:
             self.scale = 2 * math.pi
 
+        if self.emb_type == "pos" and n_mlp_layers:
+            self.mlp = MLP(self.features * 2, self.features, n_mlp_layers)
```
Comment (talmo) [inline biogtr/models/embedding.py:77]: Consider having the MLP module architecture a configurable param (e.g., `n_mlp_features`)

File: biogtr/models/global_tracking_transformer.py [hunk 6]
```diff
@@ -33,9 +32,8 @@ def __init__(
         """Initialize GTR.
 
         Args:
-            encoder_model: Name of the CNN architecture to use (e.g. "resnet18", "resnet50").
             encoder_cfg: Dictionary of arguments to pass to the CNN constructor,
-                e.g: `cfg = {"weights": "ResNet18_Weights.DEFAULT"}`
+                e.g: `cfg = {model_name: "resnet18", pretrained: False, "in_chans": 3}`
```
Comment (talmo) [inline biogtr/models/global_tracking_transformer.py:36]:
```suggestion
                e.g: `cfg = {"model_name": "resnet18", "pretrained": False, "in_chans": 3}`
```

**Review by talmo — CHANGES_REQUESTED** [review]

File: biogtr/data_structures.py (hunk 1 above)
Comment (aaprasad) [reply biogtr/data_structures.py:68]: will start doing this from now on!

File: biogtr/data_structures.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaprasad) [reply biogtr/data_structures.py:94]: will do this in a separate PR probably alongside #39

File: biogtr/models/global_tracking_transformer.py (hunk 6 above)
Comment (aaprasad) [reply biogtr/models/global_tracking_transformer.py]: done

File: biogtr/models/embedding.py (hunk 5 above)
Comment (aaprasad) [reply biogtr/models/embedding.py]: done

File: biogtr/data_structures.py (hunk 4 above)
Comment (aaprasad) [reply biogtr/data_structures.py:547]: will add this in separate PR, probably #39

**Review by talmo — APPROVED** [review]

**Pull request merged**